apps/src/weblab2/helpers/aiTutorStructuredResponseHelper.ts

```
@@ -264,55 +270,70 @@ export const getMergedAiTutorCodeWithSource = (
-      };
-      aiTutorVersionFiles.push(updatedSource.files[newFileId]);
-      return;
+    let fileId;
+    // File id is 'new' so we are creating a new file.
+    if (aiFile.id === 'new') {
+      fileId = getNextFileId(Object.values(updatedSource.files));
+    } else {
+      // File id is not 'new' so we are updating an existing file.
+      fileId = aiFile.id;
+      // We need to validate the file name of the AI code file with the given file id.
+      // If the file name of the AI code file is the same as the file name in the student code,
+      // then we use the same file id.
+      if (updatedSource.files[aiFile.id]?.name === aiFile.name) {
+        fileId = aiFile.id;
+      } else {
+        // If the file name is different, then we need to create a new file with a different id.
+        fileId = getNextFileId(Object.values(updatedSource.files));
+        // Validate the file name of the AI code file.
+        const validateFileNameError = validateFileName({
+          fileName: aiFile.name,
+          folderId: aiFile.folderId,
+          projectFiles: updatedSource.files,
+          isStartMode: false,
+          validationFile: undefined,
+        });
+        if (validateFileNameError) {
+          // If the file name is invalid (e.g. duplicate), try placing in root folder if foldierId is not already the root folder.
+          if (aiFile.folderId !== DEFAULT_FOLDER_ID) {
+            aiFile.folderId = DEFAULT_FOLDER_ID;
+            const secondValidateFileNameError = validateFileName({
+              fileName: aiFile.name,
+              folderId: aiFile.folderId,
+              projectFiles: updatedSource.files,
+              isStartMode: false,
+              validationFile: undefined,
+            });
+            if (secondValidateFileNameError) {
+              // If the 2nd attempt for the file name is invalid (e.g. duplicate), skip the AI code file update.
+              // TODO: Log error via analyticss.
```

should we add a numerical suffix instead of failing silently?

---

Ahh...I like this idea - we do this when saving to backpack. I'll update - fyi @dmcavoy -> if for some reason the id of the file that the model updates is not accurate, we create a new file instead of updating a current file, and the file name will be the name the model returns UNLESS it already exists in the folder (since folder id of an updated file is also returned by the model). If it already exists, then we create a new file in the root folder and make sure there are no duplicates by adding numerical suffix if necessary. For example, if 'dogs.html' already exists in root folder, we'll name new file 'dogs_1.html'.

---

I think adding a numerical suffix sounds great. I would rather create a file that student can delete than fail and not have them know why

---

Updated! @molly-moen - fyi, I moved logic to update `openFiles` to the helper function `getMergedAiTutorCodeWithSource` and also added unit tests. Could you give me a thumbs-up if you're good with this? Thanks!